library/src/lib/alert/alert.service.ts

```diff
@@ -1,53 +1,111 @@
+            }
+        });
+        componentRef.instance.childComponentType = content;
+
+        // Render new component
+        const domElem = (componentRef.hostView as EmbeddedViewRef<any>).rootNodes[0] as HTMLElement;
+        this.alertContainerRef.location.nativeElement.appendChild(domElem);
+
+        // Log new component
+        this.alerts.push(componentRef);
+        return alertRef;
     }
 
-    popAlert() {
-        return this.alertRef.pop();
+    public dismissAll(): void {
+        this.alerts.forEach(ref => {
+            this.destroyAlertComponent(ref);
+        });
     }
 
-    open(alertType, alertConfig?) {
-        if (typeof alertType === 'object') { // template reference variable
-            this.alertRef.push(alertType);
-        } else if (typeof alertType === 'function') { // component as content
-            const componentRef = this.componentFactoryResolver.resolveComponentFactory(alertType).create(this.injector);
-            this.alertRef.push((componentRef.instance as any).alert);
-            if (alertConfig) {
-                Object.keys(alertConfig).forEach(key => (componentRef.instance[key] = alertConfig[key]));
-            }
-            this.alertRef[this.alertRef.length - 1].instance = componentRef.instance;
-            this.appRef.attachView(componentRef.hostView);
-            const domElem = (componentRef.hostView as EmbeddedViewRef<any>).rootNodes[0] as HTMLElement;
-            document.body.appendChild(domElem);
-            const subscription = this.alertRef[this.alertRef.length - 1].afterClosed.subscribe(() => {
-                document.body.removeChild(domElem);
-                this.removeSubscription(subscription);
-            });
-            this.appRef.tick();
+    private destroyAlertComponent(alert: ComponentRef<AlertComponent>): void {
+        this.appRef.detachView(alert.hostView);
+        this.alerts[this.alerts.indexOf(alert)] = null;
+        this.alerts = this.alerts.filter(item => item !== null && item !== undefined);
+        alert.destroy();
+
+        if (!this.alerts || this.alerts.length === 0) {
+            this.destroyAlertContainer();
         }
-        this.alertRef[this.alertRef.length - 1].open();
-        return this.alertRef[this.alertRef.length - 1];
     }
 
-    removeSubscription(subscription) {
-        subscription.unsubscribe();
+    private openAlertContainer(): void {
+        const factory = this.componentFactoryResolver.resolveComponentFactory(AlertContainerComponent);
+        const componentRef = factory.create(this.injector);
+        this.appRef.attachView(componentRef.hostView);
+
+        const domElement = (componentRef.hostView as EmbeddedViewRef<any>).rootNodes[0] as HTMLElement;
+        document.body.appendChild(domElement);
+        this.alertContainerRef = componentRef;
+    }
+
+    private destroyAlertContainer(): void {
+        this.appRef.detachView(this.alertContainerRef.hostView);
```

Comment:
this will be called twice actually, is alert is manually dismissed throwing an exception with undefined `alertContainerRef`

Reply:
Not sure what you mean. In what case is this called twice?

Reply:
if you leave the default 10s timeout, but dimiss it after 2 sec, it will fire after 2, and then again after 10

Reply:
Ah I get it, nice catch. I'll just add a check in the service for an undefined container